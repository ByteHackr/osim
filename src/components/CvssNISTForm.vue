<script setup lang="ts">
import { ref } from 'vue';
import Modal from '@/components/widgets/Modal.vue';

import LabelTextarea from '@/components/widgets/LabelTextarea.vue';
import { useUserStore } from '@/stores/UserStore';

const props = defineProps<{
  cveid?: string | null | undefined;
  flawSummary?: string
  bugzilla?: string
  nvdpage?: string | null | undefined;
  cvss?: string | null;
  nistcvss?: string | null;
}>();

const userStore = useUserStore();
const show = ref(false);
const modalShown = ref(false);

const toEmail = 'nvd@nist.gov';
const ccEmail = 'secalert@redhat.com';
const subject = `CVSS Rescore Request - ${props.cveid}`;
const emailBody = `Hello,

I have performed an analysis of ${props.cveid} on behalf of Red Hat Product Security, resulting in a Red Hat CVSS score which is different from the NIST score. Our information and analysis is included below, and we would appreciate your consideration and review.

CVE : ${props.cveid}

Red Hat Bugzilla: ${props.bugzilla}
NVD Page: https://nvd.nist.gov/vuln/detail/CVE-2024-0057
Red Hat CVSS: ${props.cvss}
NIST CVSS: ${props.nistcvss}

Flaw Summary: 
${props.flawSummary}

Red Hat's CVSS Justification:
____
`;

const EmailFooter = 
`Thank you.
-${userStore.userEmail}`;

function openMailto() {
  const recipient = encodeURI('nvd@nist.gov');
  const cc = encodeURI('secalert@redhat.com');
  const encodedSubject = encodeURIComponent(subject);
  const body = emailBody + '\n\n' + EmailFooter;
  const encodedBody = encodeURIComponent(body);
  const mailto = `mailto:${recipient}?cc=${cc}&subject=${encodedSubject}&body=${encodedBody}`;
  window.open(mailto, '_blank');
}

function openModal() {
  modalShown.value = true;
}
function closeModal() {
  modalShown.value = false;
}
</script>

<template>
  <div class="osim-cve-nist-button">
    <button type="button" class="btn btn-secondary" @click="openModal">
      Email NIST
    </button>

    <Modal width="800px" :show="modalShown" @close="closeModal">
      <template #title>
        Email NIST
      </template>
      <template #body>
        <div class="osim-input mb-3 border-start ps-3">
          <p><span class="fw-bold">To:</span> <span class="to-email">{{ toEmail }}</span></p>
          <p class="from-email"><span class="fw-bold">From:</span> {{ userStore.userEmail }}</p>
          <p><span class="fw-bold">CC:</span> <span class="cc-email">{{ ccEmail }}</span></p>
          <hr />
          <p>subject:</p>
          <div class="w-100 mb-2">
            <input class="form-control" :readonly="true" type="text" :value="subject" />
          </div>
          <hr />
        </div>

        <LabelTextarea label="Body:" v-model="emailBody" />
        <LabelTextarea label="Body" v-model="EmailFooter" :hideLabel=true />
      </template>
      <template #footer>
        <button type="button" class="btn btn-secondary cancel-btn" data-bs-dismiss="modal" @click="closeModal">Cancel
        </button>
        <button type="button" @click.prevent="openMailto" class="btn btn-primary send-email">Send Email</button>
      </template>
    </Modal>
  </div>
</template>

<style scoped>
.osim-cve-nist-button :deep(.modal-dialog) {
  max-width: 800px;
}
</style>
