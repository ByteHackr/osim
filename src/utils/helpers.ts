import { toRaw, isRef, isReactive, isProxy, ref, toRef, watch } from 'vue';
import * as R from 'ramda';

export function vRef(prop: Record<string, any>, property: string, defaultValue: any) {
  const reffedProp = toRef(prop, property);
  watch(reffedProp, (value) => flexRef.value = value);
  const flexRef = reffedProp.value === undefined ? ref(defaultValue) : reffedProp;
  return flexRef;
}

export function deepToRaw<T extends Record<string, any>>(sourceObj: T): T {
  const objectIterator = (input: any): any => {
    if (Array.isArray(input)) {
      return input.map((item) => objectIterator(item));
    }
    if (isRef(input) || isReactive(input) || isProxy(input)) {
      return objectIterator(toRaw(input));
    }
    if (input && typeof input === 'object') {
      return Object.keys(input).reduce((acc, key) => {
        acc[key as keyof typeof acc] = objectIterator(input[key]);
        return acc;
      }, {} as T);
    }
    return input;
  };

  return objectIterator(sourceObj);
}

export function deepCopyFromRaw<T extends Record<string, any>>(sourceObj: T): T {
  return JSON.parse(JSON.stringify(deepToRaw(sourceObj)));
}

// eslint-disable-next-line max-len
// https://stackoverflow.com/questions/14446511/most-efficient-method-to-groupby-on-an-array-of-objects
export const groupBy = <T>(
  array: T[],
  predicate: (value: T, index: number, array: T[]) => string,
) =>
    array.reduce(
      (acc, value, index, array) => {
        (acc[predicate(value, index, array)] ||= []).push(value);
        return acc;
      },
    {} as { [key: string]: T[] },
    );

export const assignKeyValue = (object: Record<string, any>, key: string, value: any = null) => {
  object[key] = value;
  return object;
};

export const objectMap = (object: Record<string, any>, mapFn: (key: string, value: any) => any) =>
  Object.keys(object).reduce(
    (acc, key) => assignKeyValue(acc, key, mapFn(key, object[key])),
    {} as Record<string, any>,
  );

export const sortedByGroup = <T extends Record<string, any>>(array: T[], key: string) =>
  groupBy(
    array
      .filter((item: T) => item[key])
      .sort((itemA: T, itemB: T) => itemA[key].localeCompare(itemB[key])),
    (item: T) => item[key],
  );

type DeepMappable = any[] | Record<string, any>

const isNonEmptyArray = (value: any) => R.is(Array, value) && value.length > 0;
const isNonArrayObject = (value: any) => R.is(Object, value) && !R.is(Array, value);
const isDeepMappable = (value: DeepMappable) => isNonEmptyArray(value) || isNonArrayObject(value);

export const deepMap = (transform: (arg: any) => any, object: DeepMappable): any =>
  R.map(
    (val: any) => isDeepMappable(val)
      ? deepMap(transform, val)
      : transform(val)
    , object
  );

export const cveRegex = /^\s*(CVE-\d{4}-\d{4,7})\s*$/;
