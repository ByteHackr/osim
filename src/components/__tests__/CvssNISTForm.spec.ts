import { flushPromises } from '@vue/test-utils';
import { describe, it, expect } from 'vitest';

import { mountWithConfig } from '@/__tests__/helpers';

import CvssNISTForm from '../CvssNISTForm.vue';

const mountCvssNISTForm = () => mountWithConfig(CvssNISTForm, {
  props: {
    flaw: 'any',
    cveId: 'string',
    summary: 'string',
    bugzilla: 'string',
    nvdpage: 'string',
    cvss: 'string',
    nistCvss: 'string',
    cvssjustification: 'string',
  },
});

describe('cvssNISTForm', () => {
  vi.stubGlobal('open', vi.fn());

  it('renders a button', () => {
    const wrapper = mountCvssNISTForm();

    const button = wrapper.find('button');

    expect(button.exists()).toBeTruthy();
    expect(button.text()).toBe('Email NIST');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('does not render modal before clicking the button', () => {
    const wrapper = mountCvssNISTForm();

    expect(wrapper.find('div.modal-content').exists()).toBe(false);
  });

  it('renders a modal when clicking the button', async () => {
    const wrapper = mountCvssNISTForm();

    const button = wrapper.find('button');
    await button.trigger('click');
    await flushPromises();

    expect(wrapper.find('div.modal-content').exists()).toBeTruthy();
    expect(wrapper.find('input.form-control').exists()).toBeTruthy();
    expect(wrapper.find('textarea.form-control').exists()).toBeTruthy();
    expect(wrapper.find('label.osim-input').exists()).toBeTruthy();
    expect(wrapper.find('button.cancel-btn').text()).toBe('Cancel');
    expect(wrapper.find('button.send-email').text()).toBe('Send Email');
    expect(wrapper.find('span.to-email').text()).toBe('nvd@nist.gov');
    expect(wrapper.find('span.cc-email').text()).toBe('secalert@redhat.com');
    expect(wrapper.find('p.from-email').exists()).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('close modal by clicking cancel button', async () => {
    const wrapper = mountCvssNISTForm();

    const button = wrapper.find('button');
    await button.trigger('click');
    await flushPromises();

    expect(wrapper.find('div.modal-content').exists()).toBeTruthy();
    const closeButton = wrapper.find('button.cancel-btn');
    expect(closeButton.exists()).toBeTruthy();
    expect(closeButton.text()).toBe('Cancel');
    await closeButton.trigger('click');
    await flushPromises();
    expect(wrapper.find('div.modal-content').exists()).toBeFalsy();
  });

  it('open new window by clicking send button', async () => {
    const wrapper = mountCvssNISTForm();
    const button = wrapper.find('button');
    await button.trigger('click');
    await flushPromises();
    const spy = vi.spyOn(window, 'open');

    expect(wrapper.find('div.modal-content').exists()).toBeTruthy();
    const sendButton = wrapper.find('button.send-email');
    expect(sendButton.exists()).toBeTruthy();
    await sendButton.trigger('click');
    expect(spy).toHaveBeenCalledOnce();
  });
});
