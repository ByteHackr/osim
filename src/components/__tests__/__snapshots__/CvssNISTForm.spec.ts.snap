// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cvssNISTForm > renders a button 1`] = `
"<div data-v-b75fd97d="" class="osim-cve-nist-form" flaw="any" nvdpage="string" cvssjustification="string"><button data-v-b75fd97d="" type="button" class="osim-cve-email btn btn-secondary"> Email NIST </button>
  <transition-stub data-v-a347fd2c="" name="modal" appear="false" persisted="false" css="true">
    <!--v-if-->
  </transition-stub>
  <transition-stub data-v-a347fd2c="" name="modal-bg" appear="false" persisted="false" css="true">
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`cvssNISTForm > renders a modal when clicking the button 1`] = `
"<div data-v-b75fd97d="" class="osim-cve-nist-form" flaw="any" nvdpage="string" cvssjustification="string"><button data-v-b75fd97d="" type="button" class="osim-cve-email btn btn-secondary"> Email NIST </button>
  <transition-stub data-v-a347fd2c="" name="modal" appear="false" persisted="false" css="true">
    <div data-v-a347fd2c="" class="modal fade show" style="display: block;" aria-hidden="false" tabindex="-1" aria-labelledby="modalTitle" role="dialog">
      <div data-v-a347fd2c="" class="modal-dialog">
        <div data-v-a347fd2c="" class="modal-content">
          <div data-v-a347fd2c="" class="modal-header">
            <h1 data-v-a347fd2c="" class="modal-title fs-5"> Email NIST </h1><button data-v-a347fd2c="" type="button" class="btn-close" aria-label="Close"></button>
          </div>
          <div data-v-a347fd2c="" class="modal-body">
            <div data-v-b75fd97d="" class="osim-input mb-3 border-start ps-3">
              <p data-v-b75fd97d=""><span data-v-b75fd97d="" class="fw-bold">To:</span> <span data-v-b75fd97d="" class="to-email">nvd@nist.gov</span></p>
              <p data-v-b75fd97d="" class="from-email"><span data-v-b75fd97d="" class="fw-bold">From:</span> </p>
              <p data-v-b75fd97d=""><span data-v-b75fd97d="" class="fw-bold">CC:</span> <span data-v-b75fd97d="" class="cc-email">secalert@redhat.com</span></p>
              <hr data-v-b75fd97d="">
              <p data-v-b75fd97d="">Subject:</p>
              <div data-v-b75fd97d="" class="mb-2"><input data-v-b75fd97d="" class="form-control" readonly="" type="text" value="CVSS Rescore Request - string"></div>
              <hr data-v-b75fd97d="">
            </div><label data-v-1ccd17dd="" data-v-b75fd97d="" class="osim-input mb-2 ps-3">
              <div data-v-1ccd17dd="" class="row"><span data-v-1ccd17dd="" class="form-label col-3 position-relative"><!--v-if--> Body: <!--attrs: {{ $attrs }}--></span><!--    @focus="resizeTextarea"-->
                <!--    @keyup="resizeTextarea"--><textarea data-v-1ccd17dd="" class="form-control col-9 d-inline-block" rows="5" value="Hello,

I have performed an analysis of string on behalf of Red Hat Product Security,
resulting in a Red Hat CVSS score which is different from the NIST score.
Our information and analysis is included below,
and we would appreciate your consideration and review.

CVE : string

Red Hat Bugzilla: string
NVD Page: https://nvd.nist.gov/vuln/detail/string
Red Hat CVSS: string
NIST CVSS: string

Flaw Summary: 
string

Red Hat's CVSS Justification:
____

Thank you,
-
"></textarea>
                <!--v-if-->
              </div>
            </label>
          </div>
          <div data-v-a347fd2c="" class="modal-footer"><button data-v-b75fd97d="" type="button" class="btn btn-secondary cancel-btn" data-bs-dismiss="modal">Cancel </button><button data-v-b75fd97d="" type="button" class="btn btn-primary send-email"> Send Email </button></div>
        </div>
      </div>
    </div>
  </transition-stub>
  <transition-stub data-v-a347fd2c="" name="modal-bg" appear="false" persisted="false" css="true">
    <div data-v-a347fd2c="" class="modal-backdrop fade show"></div>
  </transition-stub>
</div>"
`;
